### ISSUE TITLE
Proposal: Add data binding support for Kafka module

### ISSUE THREAD

**Issue by dilanSachi** [ballerina-platform/ballerina-standard-library#2783]
## Summary
Data binding helps to access the incoming and outgoing message data in the user's desired type. Similar to the Ballerina HTTP package, subtypes of `anydata` will be the supported types. This proposal discusses ways to provide data binding for both on producer side, and the consumer side.

## Goals
- Improve user experience by adding data binding support for `kafka:Service`, `kafka:Producer` and `Kafka:Consumer`.

## Motivation
As of now, the Ballerina Kafka package does not provide direct data binding for sending and receiving messages. Only `kafka:ProducerRecord[]` and `kafka:ConsumerRecord[]` are the supported data types to send and receive messages which only support `byte[]` as the message value type. Therefore, users have to do data manipulations by themselves. With this new feature, user experience can be improved by introducing data binding to reduce the burden of developers converting byte data to the desired format as discussed in the next section.

## Description
Currently, when sending a message, user has to convert the message to `byte[]`.
```ballerina
type Person record {|
    string name;
    int age;
|};

Person person = {
    age: 10,
    name: "Bob"
};
check producer->send({ topic: TOPIC, value: person.toString().toBytes()});
```
When receiving the same message,
```ballerina
kafka:ConsumerRecord[] records = check consumer->poll(2);
foreach kafka:ConsumerRecord 'record in records {
    string messageContent = check string:fromBytes('record.value);
    Person person = check value:fromJsonStringWithType(messageContent);
}
```
Instead of this, if data binding support is introduced, users can easily send and receive the messages in the desired format.

For this purpose, we will introduce 2 new records for consuming and producing.
```ballerina
# Type related to anydata consumer record.
#
# + key - Key that is included in the record
# + value - Anydata record content
# + timestamp - Timestamp of the record, in milliseconds since epoch
# + offset - Topic partition position in which the consumed record is stored
public type AnydataConsumerRecord record {|
    anydata key?;
    anydata value;
    int timestamp;
    PartitionOffset offset?;
|};
```
```ballerina
# Details related to the anydata producer record.
#
# + topic - Topic to which the record will be appended
# + key - Key that is included in the record
# + value - Anydata record content
# + timestamp - Timestamp of the record, in milliseconds since epoch
# + partition - Partition to which the record should be sent
public type AnydataProducerRecord record {|
    string topic;
    anydata key?;
    anydata value;
    int timestamp?;
    int partition?;
|};
```
With these, user can create user-defined subtypes of the above records to achieve data binding as shown below.
```ballerina
public type PersonProducerRecord record {|
    *AnydataProducerRecord;
    string key?;
    Person value;
|};
check producer->send(personRecord);

public type PersonConsumerRecord record {|
    *AnydataConsumerRecord;
    string key?;
    Person value;
|};
PersonConsumerRecord[] personRecords = check consumer->poll(2);
```
### Listener
`kafka:Listener` receive messages in the `onConsumerRecord` method of the `kafka:Service`.
```ballerina
remote function onConsumerRecord(kafka:Caller caller, kafka:ConsumerRecord[] records) returns kafka:Error?;
```
This will be updated to accept the above-mentioned parameter types. User can create a subtype of `kafka:AnydataConsumerRecord` and use it in the function signature or directly use a subtype of `anydata` to get the payload binded directly.
Therefore, following scenarios will be available for the user.
```ballerina
remote function onConsumerRecord(kafka:Caller caller, anydata[] data, kafka:AnydataConsumerRecord[] records) returns kafka:Error?;
```
```ballerina
remote function onConsumerRecord(kafka:Caller caller, anydata[] data) returns kafka:Error?;
```
```ballerina
remote function onConsumerRecord(kafka:Caller caller, kafka:AnydataConsumerRecord[] consumerRecords) returns kafka:Error?;
```
As an example,
```ballerina
remote function onConsumerRecord(kafka:Caller caller, PersonConsumerRecord[] data) returns kafka:Error?;
remote function onConsumerRecord(kafka:Caller caller, Person[] data) returns kafka:Error?;
```
In a case where the user defined payload type is structurally same as `kafka:AnydataConsumerRecord`, user can include the `@Payload` annotation to remove the ambiguity.
```ballerina
public type RandomPayload record {|
    byte[] key?;
    byte[] value;
    int timestamp;
    record {
        record {
            string topic;
            int partition;
        } partition;
        int offset;
    } offset;
|};
remote function onConsumerRecord(kafka:AnydataConsumerRecord[] consumerRecords, @kafka:Payload RandomPayload[] payload) returns kafka:Error?;
```
### Producer
The `kafka:Producer` client has `send(kafka:ProducerRecord record)` API to send data to the Kafka server.
```ballerina
isolated remote function send(kafka:ProducerRecord producerRecord) returns Error?;
```
This will be updated as,
```ballerina
# Produces records to the Kafka server.
#
# + producerRecord - Record to be produced
# + return - A `kafka:Error` if send action fails to send data or else '()'
remote function send(kafka:AnydataProducerRecord producerRecord) returns Error?;
``` 
With this, whatever the data type given as the value will be converted to a `byte[]` internally and sent to the Kafka server. If the data binding fails, a `kafka:Error` will be returned from the API.
### Consumer
To consume messages, the consumer client has `poll()` API which returns `kafka:ConsumerRecord[]`.
```ballerina
remote function poll(decimal timeout) returns ConsumerRecord[]|Error;
```
This will be updated as follows.
```ballerina
# Polls the external broker to retrieve messages.
#
# + timeout - Polling time in seconds
# + T - Optional type description of the required data type
# + return - Array of consumer records if executed successfully or else a `kafka:Error`
isolated remote function poll(decimal timeout, typedesc<AnydataConsumerRecord[]> T = <>) returns T|Error;
```

For the cases where the user wants none of the `kafka:ConsumerRecord` information, `pollPayload()` API will be introduced.
```ballerina
isolated remote function pollPayload(decimal timeout, typedesc<anydata[]> T = <>) returns T|Error;
```
With this, user can do the following.
```ballerina
Person[] persons = check consumer->pollPayload(2);
```
With this new data binding improvement, the compiler plugin validation for `onConsumerRecord` function will also be updated to allow types of both `kafka:AnydataConsumerRecord[]` and `anydata[]`.
## Testing
- Testing the runtime data type conversions on the producer, consumer & listener.
- Testing compiler plugin validation to accept new data types.

**Comment by shafreenAnfar** [comment]
Similar to `ProducerRecord`, I think we can do the below for `ConsumerRecord`.

```ballerina
public type ConsumerRecord record {|
    byte[] key?;
    json|xml|byte[] value;
    int timestamp;
    PartitionOffset offset;
|};
```
Which means user can do the below.
```ballerina
public type Person record {|
    string name;
    string age;
|};

public type PersonRecord record {|
    *ConsumerRecord;
    Person value;
|};

PersonRecord[] persons = check consumer->poll(2);
```
If the user wants none of the `ConsumerRecord` information, we can have another API to represent that scenario. Something like the below.

```ballerina
Person[] persons = check consumer->pollWithType(2);
```
However, in both cases users have no control over the serialization method that is used. The default serialization method at the moment is json serialization.

**Comment by shafreenAnfar** [comment]
When it comes to the services what if we do the below ?

```
remote function onConsumerRecord(kafka:Caller caller, json|xml|byte[] data, kafka:ConsumerRecord[] records) returns kafka:Error?;
```

**Comment by dilanSachi** [comment]
> When it comes to the services what if we do the below ?
> 
> ```ballerina
> remote function onConsumerRecord(kafka:Caller caller, json|xml|byte[] data, kafka:ConsumerRecord[] records) returns kafka:Error?;
> ```

+1
We can also allow following scenario right?
```
remote function onConsumerRecord(kafka:Caller caller, json|xml|byte[] data) returns kafka:Error?;
```

**Comment by dilanSachi** [comment]
Updated with the suggestions

**Comment by shafreenAnfar** [comment]
LGTM!

@sameerajayasoma let us know if you have any feedback.

**Comment by dilanSachi** [comment]
We decided to allow `anydata` type instead of `json|xml|byte[]` for the intended types.
Proposal is updated to show how this will be supported.

**Comment by dilanSachi** [comment]
However, in the case of `consumer`, assume a previous code as follows,
```ballerina
kafka:ConsumerRecord[] consumerRecords = check consumer->poll(2);
byte[] value = consumerRecords[0].value;
```
If we change `byte[]` type of `kafka:ConsumerRecord` to `anydata`, this will result in a compile error and will be a breaking change.

Therefore, for now, we will be going forward with introducing `pollWithType` only.

**Comment by shafreenAnfar** [comment]
Yeah, let's not change it in that case. We can revisit it later. Besides for the 80% case, `pollWithType` should be good enough. 

Maybe later we can think of something as the below,

```ballerina
kafka:ConsumerRecord[] consumerRecords = check consumer->poll(2);
Person[] persons = kafka:bindWithType(consumerRecords);
```

**Comment by shafreenAnfar** [comment]
I've been pondering on this for a while I believe the best possible solution is as follows. It is some what similar to the origin proposal. I am thinking something as below.

Introduce two new records for consuming and producing 
```ballerina
public type AnydataConsumerRecord record {|
    anydata key?;
    anydata value;
    int timestamp?;
    PartitionOffset offset?;
|};

public type AnydataProducerRecord record {|
    string topic;
    anydata key?;
    anydata value;
    int timestamp?;
    int partition?;
|};
```
Update producer/consumer methods as follows.

## Producer
```ballerina
isolated remote function send(AnydataProducerRecord producerRecord) returns Error?
```

## Consumer
```ballerina
remote function onConsumerRecord(kafka:Caller caller, anydata[] data, kafka:AnydataConsumerRecord[] records) returns kafka:Error?;
```
```ballerina
isolated remote function poll(decimal timeout, typedesc<AnydataConsumerRecord[]> T = <>) returns T|Error;
```
```ballerina
isolated remote function pollPayload(decimal timeout, typedesc<anydata[]> T = <>) returns T|Error;
```
## Serializer/Deserializer
We will be using `anydata` serializer/deserializer for this. At the moment this is based on JSON. The serializer is engaged before producing the value and the deserializer is engaged before returning the consumed value.

This applies for both the `key` and `value`. Later we can provide a configuration to override the default `key` and `value` serializer and deserializer.

## Cosmetics 
To improve user experience and readability we can introduce `BytesConsumerRecord` and `BytesProducerRecord`. So we cover both end of the spectrum. 

```ballerina
public type BytesConsumerRecord record {|
    *AnydataConsumerRecord;
    byte[] value;
|};

public type BytesProducerRecord record {|
    *AnydataProducerRecord;
    anydata value;
|};
```  
We can deprecate the existing ones with these.

**Comment by sameerajayasoma** [comment]
Why do we need two records, such as Consumer and producer? Can we have a single record, e.g., `Message`?  That way we send and receive messages? What are the differences between producer and consumer messages?

**Comment by sameerajayasoma** [comment]
How do you differentiate between `anydata` and `kafka:AnydataConsumerRecord ` in the following remote function? By its looks, `kafka:AnydataConsumerRecord ` is a subtype of `anydata`.
```
remote function onConsumerRecord(kafka:Caller caller, anydata[] data, kafka:AnydataConsumerRecord[] records) returns kafka:Error?;
```

**Comment by shafreenAnfar** [comment]
> Why do we need two records, such as Consumer and producer? Can we have a single record, e.g., Message? That way we send and receive messages? What are the differences between producer and consumer messages?

They have their differences as follows.

```ballerina
public type AnydataConsumerRecord record {|
    anydata key?;
    anydata value;
    int timestamp?;
    PartitionOffset offset?;
|};

public type AnydataProducerRecord record {|
    anydata key?;
    anydata value;
    int timestamp?;
    int partition?;
    string topic;
|};
```
Plus one of the key design decisions we took is to make sure Ballerina messaging APIs are harmonious with each communities' APIs (Kafka, RabbitMQ, NATS/STAN). In the world of Kafka it is common to have producer and consumer records.

**Comment by shafreenAnfar** [comment]
> How do you differentiate between anydata and kafka:AnydataConsumerRecord in the following remote function? 

By looking for the most specific one (`kafka:AnydataConsumerRecord`) first.

**Comment by sameerajayasoma** [comment]
> Plus one of the key design decisions we took is to make sure Ballerina messaging APIs are harmonious with each communities' APIs (Kafka, RabbitMQ, NATS/STAN). In the world of Kafka it is common to have producer and consumer records.

Kafka Go client uses Message struct for both consumer and producer cases.

**Comment by sameerajayasoma** [comment]
> > How do you differentiate between anydata and kafka:AnydataConsumerRecord in the following remote function?
> 
> By looking for the most specific one (`kafka:AnydataConsumerRecord`) first.

AFAIR, we faced the same situation in HTTP resource function return types. That's why we had to create `http:Ok`  record such that its not a subtype of anydata, IIRC

**Comment by shafreenAnfar** [comment]
> Kafka Go client uses Message struct for both consumer and producer cases. 

True, we were thinking Java as the most prominent client in the community (well, maybe we were biased :)). However, the main reason for having two structures is not that but having different information in addition to overlapping once.

**Comment by shafreenAnfar** [comment]
> AFAIR, we faced the same situation in HTTP resource function return types. That's why we had to create http:Ok record such that its not a subtype of anydata, IIRC

True, if the user has a payload similar to consumer record, there is no way to distinguish both scenarios. In addition to that, we had another concerns about this, which is `onConsumerRecord` doing two different things.

1) Binding subtypes of `AnydataConsumerRecord`
2) Binding to the payload of `AnydataConsumerRecord`

**Note**: Only first is implemented. 

However, given that we already have the problem @sameerajayasoma pointed out. I think only (and the right) option is to introduce a new `remote` method named `onConsumerPayload`.

**Comment by shafreenAnfar** [comment]
@sameerajayasoma, @dilanSachi and I had a chat on this. In addition to the above following were key points.

- Should `pollPayload` be `pollValue`
   - I think in the context of Kafka `pollValue` makes sense too. But `payload` has a well-known meaning [1]. Therefore, I feel that latter is better.
- Should `ConsumerRecord` and `ProducerRecord` be `ConsumerMessage` and `ProducerMessage`. Here the concern was usage of the word `record`
   - I think both terms are fine [2]. Therefore, I feel we should stick to `record` as we have already used it. 

@sameerajayasoma @dilanSachi any thoughts ?

[1] https://en.wikipedia.org/wiki/Payload_(computing)
[2] https://kafka.apache.org/documentation/#intro_concepts_and_terms

**Comment by shafreenAnfar** [comment]
@sameerajayasoma and I had a chat about the above comments and agreed on the decision points mentioned in the comment. 

However, with regard to introducing `onConsumerPayload` remote method as suggested in https://github.com/ballerina-platform/ballerina-standard-library/issues/2783#issuecomment-1111938729, we decided against it. Instead, we decided to introduce an annotation named `@Payload` as in HTTP package.  When there is an ambiguity between `AnydataConsumerRecord` and the actual payload, users can use the proposed annotation and resolve the ambiguity. Just to cement the point this annotation is only needed when there is an ambiguity. For example, for some reason if the payload is a subtype of `AnydataConsumerRecord`. In all the other cases this is not needed. Therefore, if you apply 80/20 rule, for the 80% case users do not need to use this annotation.
